data-prepper-core/src/main/java/com/amazon/dataprepper/peerforwarder/PeerForwarderReceiveBuffer.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.peerforwarder;
+
+import com.amazon.dataprepper.model.CheckpointState;
+import com.amazon.dataprepper.model.buffer.Buffer;
+import com.amazon.dataprepper.model.buffer.SizeOverflowException;
+import com.amazon.dataprepper.model.record.Record;
+import com.google.common.base.Stopwatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static java.lang.String.format;
+
+/**
+ * Buffer created for each stateful processor which implements {@link com.amazon.dataprepper.model.peerforwarder.RequiresPeerForwarding}
+ * interface.
+ *
+ * @since 2.0
+ */
+public class PeerForwarderReceiveBuffer<T extends Record<?>> implements Buffer<T> {
```

Comment:
This class is nearly identical to the BlockingBuffer. Let's create a backlog item to address duplicate code.